File: ios/BFExplorer/Plaoc/bfs/ipc/Ipc.swift
//
//  Ipc.swift
//  BFExplorer
//
//  Created by ui08 on 2023/2/8.
//

import Foundation
import Vapor
import Combine

var ipc_uid_acc = 0
class Ipc {
    typealias IpcTupleCtor = () -> SIGNAL_CTOR
    typealias IpcTupleBool = () -> Bool
    
    var uid: Int = ipc_uid_acc++
    /**
     * 是否支持 messagePack 协议传输：
     * 需要同时满足两个条件：通道支持直接传输二进制；通达支持 MessagePack 的编解码
     */
    var support_message_pack: Bool
    /**
     * 是否支持 Protobuf 协议传输：
     * 需要同时满足两个条件：通道支持直接传输二进制；通达支持 Protobuf 的编解码
     */
    var support_protobuf: Bool
    /** 是否支持 二进制 传输 */
    var suport_bianry: Bool {
        get {
            return support_message_pack || support_protobuf
        }
    }
    var remote: MicroModule
    var role: IPC_ROLE
    
    init() {
        support_message_pack = false
        support_protobuf = false
        remote = NativeMicroModule()
        role = IPC_ROLE.client
//        onMessage = _messageSignal.listen
    }

    internal var _messageSignal = Signal<(IpcMessage, Ipc)>()
    func postMessage(message: IpcMessage) -> Void {
        if self._closed {
            return
        }
        
        self._doPostMessage(data: message)
    }
    
//    var onMessage: ((@escaping OnIpcMessage) -> IpcTupleBool)
    func onMessage(cb: @escaping OnIpcMessage) -> IpcTupleBool {
        return _messageSignal.listen(cb)
    }
    func _doPostMessage(data: IpcMessage) {}
    
    private lazy var _getOnRequestListener = {
        let signal = Signal<(IpcMessage, Ipc)>()
        _ = _messageSignal.listen { (message, ipc) in
            if let message = message as? IpcRequest {
                signal.emit((message, ipc))
            }
        }

        return signal.listen
    }()
    
    
    func onRequest(cb: @escaping ((IpcMessage, Ipc)) -> SIGNAL_CTOR) -> IpcTupleBool {
        return _getOnRequestListener(cb)
    }
    
    private var _closed = false
    private var _closeSignal = Signal<()>()
    
    func _doClose() async {}
    
    func close() async {
        if self._closed {
            return
        }
        
        self._closed = true
        await self._doClose()
        self._closeSignal.emit(())
    }
    
    func onClose(cb: @escaping IpcTupleCtor) -> IpcTupleBool {
        return self._closeSignal.listen(cb)
    }
    
    func request(request: Request) async -> IpcResponse {
        return await withCheckedContinuation { continuation in
            var req_id = allocReqId()
            
            self.postMessage(message: IpcRequest.fromRequest(req_id: req_id, request: request, ipc: self))
            
            _ = self.onMessage { (message, ipc) in
                if let message = message as? IpcResponse, req_id == message.req_id {
                    continuation.resume(returning: message)
                }
            }
        }
    }
    private var _req_id_acc = 0
    func allocReqId() -> Int {
        return self._req_id_acc++
    }
    
}

File: android/app/src/main/java/info/bagen/rust/plaoc/microService/ipc/Ipc.kt
package info.bagen.rust.plaoc.microService.ipc

import info.bagen.rust.plaoc.microService.ipc.helper.*

var ipc_uid_acc = 0

abstract class Ipc {
    abstract val supportMessagePack: Boolean
    val uid = ipc_uid_acc++
    val remote:TMicroModule  get() { return  TMicroModule() }
    val role: IPC_ROLE get() { return IPC_ROLE.SERVER }

    val _messageSignal = createSignal<ipcSignal>()
    fun postMessage(message: IpcMessage) {
        if (this._closed) {
            return;
        }
        this._doPostMessage(message);
    }

    abstract fun _doPostMessage(data: IpcMessage): Void;
    private fun _getOnRequestListener(cb: OnIpcRequestMessage): Signal<ipcSignal> {
        val signal = createSignal<ipcSignal>()
        this._messageSignal.listen { args ->
            val request = args[0] as IpcRequest
            val ipc = args[1]  as Ipc
            if (IPC_DATA_TYPE.REQUEST == request.type) {
                signal.emit(request, ipc);
            }
        }
        return signal
    }

   fun onRequest(cb: OnIpcRequestMessage): Signal<ipcSignal> {
        return this._getOnRequestListener(cb);
    }
    abstract fun _doClose(): Void

    private var _closed = false
    fun close() {
        if (this._closed) {
            return;
        }
        this._closed = true;
        this._doClose();
        this._closeSignal.emit(null,null)
    }
    private val _closeSignal = createSignal<ipcSignal>();
    val onMessage = this._messageSignal
    val onClose =  this._closeSignal

    private val _reqresMap = mutableMapOf<Number, IpcResponse>();
    private var _req_id_acc = 0;
    fun allocReqId(): Int {
        return this._req_id_acc++;
    }

    private var _inited_req_res = false;
    private fun _initReqRes() {
        if (this._inited_req_res) {
            return;
        }
        this._inited_req_res = true;
//        this._messageSignal.listen(fun(message){
//            if (IPC_DATA_TYPE.RESPONSE.equals(message.type)) {
//                var response_po = this._reqresMap.get(message.req_id);
//                if (response_po) {
//                    this._reqresMap.delete(message.req_id);
//                    response_po.resolve(message);
//                } else {
//                    throw new Error(`no found response by req_id: ${message.req_id}`);
//                }
//            }
//        });
    }
}


